### PR TITLE
fix: order of DKIM/DMARC milters matters

### PR DESCRIPTION
**Pull request by casperklein**
# Description

This PR fixes a regression introduced with #3016, where the postfix setting `smtpd_milters ` changed the order of the values 
from `smtpd_milters = $dkim_milter $dmarc_milter`
to `smtpd_milters = $dmarc_milter $dkim_milter`.

The changed order led to this kind of DMARC failures:

<details>
<summary>Details</summary>

```
Jan 31 21:00:05 mail postfix/postscreen[22649]: CONNECT from [209.85.128.51]:53204 to [172.19.0.2]:25
Jan 31 21:00:11 mail postfix/postscreen[22649]: PASS NEW [209.85.128.51]:53204
Jan 31 21:00:11 mail postfix/smtpd[22650]: connect from mail-wm1-f51.google.com[209.85.128.51]
Jan 31 21:00:11 mail postfix/smtpd[22650]: Anonymous TLS connection established from mail-wm1-f51.google.com[209.85.128.51]: TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits) key-exchange X25519 server-signature RSA-PSS (2048 bits) server-digest SHA256
Jan 31 21:00:11 mail postfix/smtpd[22650]: warning: restriction `reject_authenticated_sender_login_mismatch' ignored: no SASL support
Jan 31 21:00:11 mail policyd-spf[22655]: prepend Received-SPF: Pass (mailfrom) identity=mailfrom; client-ip=209.85.128.51; helo=mail-wm1-f51.google.com; envelope-from=foobar+caf_=postmaster=mydomain.tld@gmail.com; receiver=<UNKNOWN>
Jan 31 21:00:11 mail postfix/smtpd[22650]: A9F9335803BE: client=mail-wm1-f51.google.com[209.85.128.51]
Jan 31 21:00:11 mail postfix/cleanup[22656]: A9F9335803BE: message-id=<00000000000065e72305f393f049@google.com>
Jan 31 21:00:11 mail opendmarc[2523]: A9F9335803BE ignoring Authentication-Results at 15 from mx.google.com
Jan 31 21:00:11 mail opendmarc[2523]: A9F9335803BE: google.com fail
Jan 31 21:00:11 mail postfix/cleanup[22656]: A9F9335803BE: milter-reject: END-OF-MESSAGE from mail-wm1-f51.google.com[209.85.128.51]: 5.7.1 rejected by DMARC policy for google.com; from=<foobar+caf_=postmaster=mydomain.tld@gmail.com> to=<postmaster@mydomain.tld> proto=ESMTP helo=<mail-wm1-f51.google.com>
Jan 31 21:00:44 mail postfix/smtpd[22650]: disconnect from mail-wm1-f51.google.com[209.85.128.51] ehlo=2 starttls=1 mail=1 rcpt=1 bdat=0/1 rset=1 quit=1 commands=7/8
```
</details>

I switched back the order of the two conditions:

1. `[[ ${ENABLE_OPENDMARC} -eq 1 ]]`
2. `[[ ${ENABLE_OPENDKIM} -eq 1 ]]`

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ap-wtioit** [comment]
As mentioned in #3078 this fix looks good to me (and necessary), but it doesn't fix #3078.

It should fix the issue that opendmarc doesn't take into account the DKIM signatures into the DMARC check. (Which is an issue that also happens on our servers)

**Review by ap-wtioit — APPROVED** [review]

File: target/scripts/startup/setup-stack.sh [hunk 1]
```diff
@@ -634,51 +634,52 @@ function _setup_SRS
 
 function _setup_dkim_dmarc
 {
-  if [[ ${ENABLE_OPENDMARC} -eq 1 ]]
+  if [[ ${ENABLE_OPENDKIM} -eq 1 ]]
```
Comment (ap-wtioit) [inline target/scripts/startup/setup-stack.sh:637]: it works for our servers (opendkim + opendmarc enabled) but the `return 0` is lost here.

which syntactically was useless in the current state of the code (after lot's of refactoring) but made me think why it was there in the first place.

should opendmarc be enabled without opendkim (and policyd-spf) or should there be at least a warning that all mails for domains with DMARC enable are incorrectly rejected if you run opendmarc without policyd-spf or without opendkim?

for a [RFC 7489](https://www.rfc-editor.org/rfc/rfc7489#page-24) compliant mail receiver setup if opendmarc is enabled (ENABLE_OPENDMARC), policyd-spf (always enabled) and opendkim (ENABLE_OPENDKIM) must be enabled.

Comment (casperklein) [reply target/scripts/startup/setup-stack.sh:637]: > should opendmarc be enabled without opendkim (and policyd-spf) or should there be at least a warning that all mails for domains with DMARC enable are incorrectly rejected if you run opendmarc without policyd-spf or without opendkim?

There is definitely room for improvments. This PR only aims to fix the current regression. At the moment, I don't have much spare time. So looking into it, will take same weeks.

**Review by polarathene — APPROVED** [review]
I think this will conflict with a related PR by @georglauterbach 

One could also probably just append both milter values to an array and afterwards instead of similar logic for both, just have one edit after both dkim + dmarc options are handled to add enabled milters. That way the existing dmarc first would still keep dmarc first in the array, then dkim (_probably worth noting order is important with a comment though_).

**Comment by georglauterbach** [comment]
Yeah, I had actually hoped to get my PR merged before, but I'm lacking the reviews 😅

I can take care of the merge conflicts, but I'd really love to merge my PRs then :)

**Review by georglauterbach — APPROVED** [review]

**Comment by polarathene** [comment]
> I can take care of the merge conflicts

You could probably just squeeze in the fix in that PR technically and we could close this one.

No need to swap the order like this PR is doing with moving conditional blocks. Just append to an array and after the two update the config like I said? (_although I recall you changing opendkim conditional into an early return guard instead, which I wasn't a fan of to skip some indentation since it wasn't consistent with the related dmarc conditional_ :sweat_smile: )

**Comment by georglauterbach** [comment]
> > I can take care of the merge conflicts
>
> You could probably just squeeze in the fix in that PR technically and we could close this one.

I will just go ahead and merge this one so it's out on `master`.

> No need to swap the order like this PR is doing with moving conditional blocks. Just append to an array and after the two update the config like I said? (_although I recall you changing opendkim conditional into an early return guard instead, which I wasn't a fan of to skip some indentation since it wasn't consistent with the related dmarc conditional_ :sweat_smile: )

I will resolve the conflicts and adapt your strategy :) This way the work @casperklein has put into the PR is not wasted IMO :)

**Pull request merged**